R/trace_tests.R

```diff
@@ -213,38 +217,100 @@ update_current_test <- function() {
   .current_test$srcref <- getSrcref(.current_test$src_call)
   .current_test$src <- .current_test$srcref %||% .current_test$src_call
 
-  # build test data to store within .counters
-  test <- list(.current_test$trace)
-
-  # only name if srcrefs can be determined
-  if (inherits(.current_test$src, "srcref")) {
-    names(test) <- file.path(
-      dirname(get_source_filename(.current_test$src, normalize = TRUE)),
-      key(.current_test$src))
-  }
+  .current_test$key <- current_test_key()
+  .current_test$index <- current_test_index()
+  .current_test$call_count <- current_test_call_count()
 
   # NOTE: r-bugs 18348
   # restrict test call lengths to avoid R Rds deserialization limit
   # https://bugs.r-project.org/show_bug.cgi?id=18348
   max_call_len <- 1e4
-  call_lengths <- vapply(test[[1L]], length, numeric(1L))
+  call_lengths <- vapply(.current_test$trace, length, numeric(1L))
   if (any(call_lengths > max_call_len)) {
-    test[[1L]] <- lapply(test[[1L]], truncate_call, limit = max_call_len)
+    .current_test$trace <- lapply(
+      .current_test$trace,
+      truncate_call,
+      limit = max_call_len
+    )
+
     warning("A large call was captured as part of a test and will be truncated.")
   }
 
-  .counters$tests <- append(.counters$tests, test)
+  .counters$tests[[.current_test$index]] <- .current_test$trace
+  attr(.counters$tests[[.current_test$index]], "call_count") <- .current_test$call_count
+  names(.counters$tests)[[.current_test$index]] <- .current_test$key
+}
+
+#' Build key for the current test
+#'
+#' If the current test has a srcref, a unique character key is built from its
+#' srcref. Otherwise, an empty string is returned.
+#'
+#' @return A unique character string if the test call has a srcref, or an empty
+#'   string otherwise.
+#'
+#' @keywords internal
+current_test_key <- function() {
+  if (!inherits(.current_test$src, "srcref")) return("")
+  file.path(
+    dirname(get_source_filename(.current_test$src, normalize = TRUE)),
+    key(.current_test$src)
+  )
 }
 
+#' Retrieve the index for the test in `.counters$tests`
+#'
+#' If the test was encountered before, the index will be the index of the test
+#' in the logged tests list. Otherwise, the index will be the next index beyond
+#' the length of the tests list.
+#'
+#' @return An integer index for the test call
+#'
+#' @keywords internal
+current_test_index <- function() {
+  # check if test has already been encountered and reuse test index
+  if (inherits(.current_test$src, "srcref")) {
+    # when tests have srcrefs, we can quickly compare test keys
+    match(
+      .current_test$key,
+      names(.counters$tests),
+      nomatch = length(.counters$tests) + 1L
+    )
+  } else {
+    # otherwise we compare call stacks
+    Position(
+      function(t) identical(t[], .current_test$trace),  # t[] to ignore attr
+      .counters$tests,
+      right = TRUE,
+      nomatch = length(.counters$tests) + 1L
+    )
+  }
+}
```

Comment:
If the current test matches an already logged test, then we reuse that index. Otherwise it's added to a growing list of recorded test expressions.

New tests are distinguished by comparing the srcref "key" to any previous test keys if the test code has known srcrefs, or otherwise looking for identical call stacks.

I was expecting call stack comparisons like this to come with a big performance hit, but it ended up being pretty minor.